File: src/pythonnative/native/android/src/main/java/com/pythonnative/runtime/PNBridge.kt
package com.pythonnative.runtime

import android.app.Activity
import android.content.Context
import android.util.Log
import com.pythonnative.runtime.animation.PNAnimator
import com.pythonnative.runtime.bridge.JsonUtil
import com.pythonnative.runtime.bridge.MainThread
import com.pythonnative.runtime.bridge.PNRegistry
import com.pythonnative.runtime.bridge.TransactionApplier
import com.pythonnative.runtime.bridge.ViewRegistry
import com.pythonnative.runtime.modules.BuiltinModules
import com.pythonnative.runtime.modules.ModuleEnvelope
import com.pythonnative.runtime.modules.Promise
import org.json.JSONObject
import java.util.Locale

/**
 * The single Java class Python touches.
 *
 * Every entry point mirrors one C symbol of the iOS bridge and is
 * expected to run on the main thread (Python's asyncio loop lives
 * there). Calls from other threads are logged but still executed so a
 * misbehaving caller fails loudly in logcat instead of silently.
 */
object PNBridge {
    /** Logcat tag used by the whole runtime. */
    const val TAG = "PythonNative"

    /** Protocol version compiled into this library. */
    const val PROTOCOL_VERSION = 2

    private val pythonQueue = java.util.concurrent.Executors.newSingleThreadExecutor { runnable ->
        Thread(runnable, "PythonNative-events").apply { isDaemon = true }
    }
    @Volatile private var host: PythonHost? = null
    private var activity: Activity? = null
    private val hostReadyListeners = ArrayList<Runnable>()
    private val modulePromises = java.util.concurrent.ConcurrentHashMap<Long, Promise>()

    /** Tag to view bookkeeping shared by the applier and managers. */
    val registry = ViewRegistry()

    private data class Message(val kind: String, val tag: Long, val name: String, val payload: String) {
        val continuous get() = kind == "event" && (name in setOf("on_scroll", "on_selection_change", "on_gesture_update") ||
            name.startsWith("gesture:") && JSONObject(payload).optJSONArray("args")?.optJSONObject(0)?.optString("state") == "changed")
    }
    private val mailbox = java.util.ArrayDeque<Message>()
    private var draining = false

    private val commits = com.pythonnative.runtime.bridge.CommitState()
    private val applier by lazy { TransactionApplier(registry) }

    /** The protocol version Python must match before starting. */
    @JvmStatic
    fun protocolVersion(): Int = PROTOCOL_VERSION

    /** Install the Python callback target; flushes callbacks queued before it existed. */
    @JvmStatic
    fun setHost(host: PythonHost) {
        this.host = host
        val pending = ArrayList(hostReadyListeners)
        hostReadyListeners.clear()
        for (listener in pending) {
            try {
                listener.run()
            } catch (e: Exception) {
                Log.e(TAG, "host-ready listener failed", e)
            }
        }
    }

    /** Whether a [PythonHost] has been installed. */
    fun hasHost(): Boolean = host != null

    /** Run `runnable` now if a host is installed, otherwise once [setHost] is called. */
    fun runWhenHostReady(runnable: Runnable) {
        if (host != null) runnable.run() else hostReadyListeners.add(runnable)
    }

    /** Record the current activity; modules and managers derive their `Context` from it. */
    @JvmStatic
    fun setContext(activity: Activity) {
        this.activity = activity
        PNRegistry.ensureBuiltins()
        BuiltinModules.onContextAttached(activity)
    }

    /** Forget `activity` if it is the current one (called from `Activity.onDestroy`). */
    @JvmStatic
    fun clearContext(activity: Activity) {
        if (this.activity === activity) this.activity = null
    }

    /** The current activity, or `null` before [setContext]. */
    fun activity(): Activity? = activity

    /** The context used to create views; throws before [setContext]. */
    fun context(): Context =
        activity ?: throw IllegalStateException("PNBridge.setContext(activity) has not been called")

    /** Display density of the current context (1.0 when no context is attached). */
    fun density(): Float {
        val ctx = activity ?: return 1f
        val d = ctx.resources.displayMetrics.density
        return if (d > 0f) d else 1f
    }

    // ------------------------------------------------------------------
    // Python -> native
    // ------------------------------------------------------------------

    /** Apply one serialized transaction (a JSON array of ops). */
    @JvmStatic
    fun apply(transactionJson: String): String {
        if (!MainThread.isMain()) return MainThread.call { apply(transactionJson) }
        return commits.apply(transactionJson, applier)
    }

    /**
     * Measure the view with `tag` under the given constraints (dp; `1e6`
     * means unconstrained) and return `"w,h"` in dp.
     */
    @JvmStatic
    fun measure(tag: Long, maxWidth: Double, maxHeight: Double): String {
        if (!MainThread.isMain()) return MainThread.call { measure(tag, maxWidth, maxHeight) }
        val record = registry.get(tag)
        if (record == null) {
            Log.w(TAG, "measure: unknown tag $tag")
            return "0,0"
        }
        val size = try {
            record.manager.measure(record.view, maxWidth, maxHeight)
        } catch (e: Exception) {
            Log.e(TAG, "measure failed for ${record.typeName}#$tag", e)
            floatArrayOf(0f, 0f)
        }
        return String.format(Locale.US, "%.3f,%.3f", size[0], size[1])
    }

    /** Run an imperative command on one view and return its JSON result. */
    @JvmStatic
    fun command(tag: Long, name: String, argsJson: String): String? {
        if (!MainThread.isMain()) return MainThread.call { command(tag, name, argsJson) }
        val record = registry.get(tag) ?: return null
        return try {
            val args = if (argsJson.isBlank()) JSONObject() else JSONObject(argsJson)
            JsonUtil.encode(record.manager.command(record.view, name, args))
        } catch (e: Exception) {
            Log.e(TAG, "command '$name' failed for ${record.typeName}#$tag", e)
            null
        }
    }

    /** Handle an animation request (`set`, `start`, or `cancel`) for one view. */
    @JvmStatic
    fun animate(tag: Long, requestJson: String): String? {
        if (!MainThread.isMain()) return MainThread.call { animate(tag, requestJson) }
        val record = registry.get(tag) ?: return null
        return try {
            PNAnimator.handle(record, JSONObject(requestJson))
        } catch (e: Exception) {
            Log.e(TAG, "animate failed for ${record.typeName}#$tag: $requestJson", e)
            null
        }
    }

    /** Call a native module method; `argsJson` is `{"call_id": n, "args": {...}}`. */
    @JvmStatic
    fun call(module: String, method: String, argsJson: String): String? {
        if (!MainThread.isMain()) return MainThread.call { call(module, method, argsJson) }
        val (callId, args) = try {
            ModuleEnvelope.decodeCall(argsJson)
        } catch (e: Exception) {
            return ModuleEnvelope.error("malformed call envelope: ${e.message}", "bad_request")
        }
        if (method == "_pn_cancel") {
            modulePromises.remove(args.optLong("call_id"))?.takeIf { it.module == module }?.cancel()
            return ModuleEnvelope.ok(null)
        }
        if (module == "Runtime") return JSONObject().put("ok", true).put("value", JSONObject()
            .put("protocol", 2).put("yoga", "3.2.1").put("schema", com.pythonnative.generated.PNContracts.fingerprint)
            .put("animation_graph", true).put("logical_lists", true).put("native_layout", true)).toString()
        if (module == "Layout") return JSONObject().put("ok", true).put("value",
            com.pythonnative.runtime.layout.NativeLayout.compute(args)).toString()
        val target = PNRegistry.module(module)
            ?: return ModuleEnvelope.error("unknown module '$module'", "unknown_module")
        val promise = Promise(callId, module)
        if (callId > 0) {
            modulePromises[callId] = promise
            promise.onFinished = { modulePromises.remove(callId) }
        }
        try {
            target.call(method, args, promise)
        } catch (e: Exception) {
            Log.e(TAG, "module $module.$method threw", e)
            if (!promise.isSettled) promise.reject(e.message ?: e.toString(), "exception")
        }
        return promise.markReturned()
    }

    // ------------------------------------------------------------------
    // native -> Python
    // ------------------------------------------------------------------

    /**
     * Deliver one callback to Python. Returns `null` when no host is
     * installed or the Python handler raised (the error is logged).
     */
    fun callPython(kind: String, tag: Long, name: String, payloadJson: String): String? {
        val target = host
        if (target == null) {
            Log.w(TAG, "callPython($kind, $tag, $name) dropped: no PythonHost installed")
            return null
        }
        if (kind == "event") com.pythonnative.runtime.animation.AnimationGraph.event(tag, name, org.json.JSONArray(payloadJson))
        var editRevision = 0L
        if (kind == "event" && name == "on_change") registry.get(tag)?.let { record ->
            if (record.typeName == "TextInput") {
                editRevision = ((record.state["edit_revision"] as? Number)?.toLong() ?: 0L) + 1L
                record.state["edit_revision"] = editRevision
            }
        }
        val payload = if (kind == "event") commits.event(org.json.JSONArray(payloadJson), editRevision) else payloadJson
        val message = Message(kind, tag, name, payload)
        val schedule = synchronized(mailbox) {
            val last = mailbox.peekLast()
            if (message.continuous && last?.continuous == true && last.tag == tag && last.name == name) mailbox.removeLast()
            mailbox.addLast(message)
            val needed = !draining
            draining = true
            needed
        }
        if (schedule) pythonQueue.execute {
            while (true) {
                val next = synchronized(mailbox) {
                    mailbox.pollFirst().also { if (it == null) draining = false }
                } ?: break
                try { target.callback(next.kind, next.tag, next.name, next.payload) }
                catch (error: Exception) { Log.e(TAG, "Python callback ${next.kind}/${next.name} raised", error) }
            }
        }
        return null
    }

    // ------------------------------------------------------------------
    // Threading helpers
    // ------------------------------------------------------------------

    /** Whether the caller is on the main thread. */
    fun isMainThread(): Boolean = MainThread.isMain()

    /** Post `runnable` to the main thread. */
    fun post(runnable: Runnable) = MainThread.post(runnable)

    private fun assertMain(entry: String) {
        if (!MainThread.isMain()) {
            Log.w(TAG, "PNBridge.$entry called off the main thread (${Thread.currentThread().name})")
        }
    }
}

File: src/pythonnative/native/ios/Sources/PythonNativeKit/Bridge/PNBridge.swift
import Foundation
import UIKit

/// The version of the wire protocol compiled into this library. Python
/// refuses to start when `pythonnative.bridge.PROTOCOL_VERSION` differs.
public let PNProtocolVersion: Int32 = 2

/// The C signature Python registers through `pn_bridge_set_callback`.
///
/// `(kind, tag, name, payload_json) -> optional JSON string owned by
/// Python and valid until the next callback returns`.
public typealias PNCallbackFn = @convention(c) (
    UnsafePointer<CChar>?, Int64, UnsafePointer<CChar>?, UnsafePointer<CChar>?
) -> UnsafePointer<CChar>?

/// Swift-side entry point for the reverse direction (native -> Python) and
/// the owner of the registered C callback.
public final class PNBridge {
    public static let shared = PNBridge()

    private let pythonQueue = DispatchQueue(label: "dev.pythonnative.events")
    private struct Message {
        let kind: String; let tag: Int64; let name: String; let payload: String
        var continuous: Bool {
            guard kind == "event" else { return false }
            if ["on_scroll", "on_selection_change", "on_gesture_update"].contains(name) { return true }
            if name.hasPrefix("gesture:"), let args = PNJSON.decodeObject(payload)["args"] as? [[String: Any]] {
                return args.first?["state"] as? String == "changed"
            }
            return false
        }
    }
    private let mailboxLock = NSLock()
    private var mailbox: [Message] = []
    private var draining = false
    private var callback: PNCallbackFn?
    private var registrationWaiters: [() -> Void] = []

    private init() {
        // Reference every exported entry point so the linker never
        // dead-strips them out of the host executable; Python resolves
        // them through `dlsym` at runtime.
        _ = PNBridge.exportedSymbols
    }

    /// Whether Python has registered its callback yet.
    public var hasCallback: Bool { callback != nil }

    /// Install (or clear) the Python callback. Runs any pending
    /// `whenCallbackRegistered` blocks when a callback becomes available.
    public func setCallback(_ fn: PNCallbackFn?) {
        callback = fn
        guard fn != nil else { return }
        let waiters = registrationWaiters
        registrationWaiters = []
        for waiter in waiters { waiter() }
    }

    /// Run `block` immediately if the callback is registered, otherwise
    /// once it is. Used to buffer deep links and other early events.
    public func whenCallbackRegistered(_ block: @escaping () -> Void) {
        if callback != nil {
            block()
        } else {
            registrationWaiters.append(block)
        }
    }

    /// Invoke the Python callback. Returns the JSON string Python returned
    /// (copied immediately), or `nil` when Python returned NULL or no
    /// callback is registered.
    @discardableResult
    public func callPython(kind: String, tag: Int64, name: String, payload: String) -> String? {
        let message = Message(kind: kind, tag: tag, name: name, payload: payload)
        mailboxLock.lock()
        if message.continuous, let last = mailbox.last, last.continuous,
           last.tag == tag, last.name == name {
            mailbox[mailbox.count - 1] = message
        } else { mailbox.append(message) }
        let schedule = !draining
        draining = true
        mailboxLock.unlock()
        if schedule { pythonQueue.async { [self] in drainMailbox() } }
        return nil
    }

    private func drainMailbox() {
        while true {
            mailboxLock.lock()
            if mailbox.isEmpty { draining = false; mailboxLock.unlock(); return }
            let message = mailbox.removeFirst()
            mailboxLock.unlock()
            guard let callback = callback else { continue }
            message.kind.withCString { kind in
                message.name.withCString { name in
                    message.payload.withCString { payload in
                        _ = callback(kind, message.tag, name, payload)
                    }
                }
            }
        }
    }

    static func onUI<T>(_ body: () -> T) -> T {
        if Thread.isMainThread { return body() }
        return DispatchQueue.main.sync(execute: body)
    }

    /// Emit a view event: `callback("event", tag, name, args_array_json)`.
    @discardableResult
    public func emitEvent(tag: Int64, name: String, args: [Any?]) -> String? {
        PNAnimationGraph.event(tag, name, args)
        var editRevision = 0
        if name == "on_change", let view = PNViewRegistry.shared.view(for: tag),
           view is UITextInput, let state = PNViewState.existing(for: view) {
            editRevision = (state.extras["edit_revision"] as? Int ?? 0) + 1
            state.extras["edit_revision"] = editRevision
        }
        return callPython(kind: "event", tag: tag, name: name, payload: PNCommit.event(args, editRevision: editRevision))
    }

    /// Log when an entry point is reached off the main thread. Python
    /// always calls from the main thread, so this only guards misuse.
    static func ensureMainThread(_ context: String) {
        if !Thread.isMainThread {
            PNLog.rateLimited(PNLog.bridge, key: "off-main:\(context)", "\(context) called off the main thread; running inline")
        }
    }

    /// Copy a Swift string into a `strdup`'d buffer Python frees via `pn_bridge_free`.
    static func duplicate(_ string: String?) -> UnsafeMutablePointer<CChar>? {
        guard let string = string else { return nil }
        return strdup(string)
    }

    private static let exportedSymbols: [Any] = [
        pn_bridge_apply as Any,
        pn_bridge_measure as Any,
        pn_bridge_command as Any,
        pn_bridge_animate as Any,
        pn_bridge_call as Any,
        pn_bridge_free as Any,
        pn_bridge_set_callback as Any,
        pn_bridge_protocol_version as Any,
    ]
}

// MARK: - C exports

/// Apply one serialized transaction (a JSON array of ops).
@_cdecl("pn_bridge_apply")
public func pn_bridge_apply(_ transactionJSON: UnsafePointer<CChar>?) -> UnsafeMutablePointer<CChar>? {
    guard let transactionJSON = transactionJSON else { return nil }
    let json = String(cString: transactionJSON)
    return PNBridge.onUI { PNBridge.duplicate(PNCommit.apply(json)) }
}

/// Measure the view with `tag` under the given constraints (`1e6` means unconstrained).
@_cdecl("pn_bridge_measure")
public func pn_bridge_measure(
    _ tag: Int64, _ maxW: Double, _ maxH: Double,
    _ outW: UnsafeMutablePointer<Double>?, _ outH: UnsafeMutablePointer<Double>?
) {
    PNBridge.onUI {

    PNBridge.ensureMainThread("pn_bridge_measure")
    var size = CGSize.zero
    if let record = PNViewRegistry.shared.resolve(tag) {
        size = record.manager.measure(view: record.view, maxW: CGFloat(maxW), maxH: CGFloat(maxH))
    }
    outW?.pointee = size.width.isFinite ? Double(size.width) : 0
    outH?.pointee = size.height.isFinite ? Double(size.height) : 0
    }
}

/// Run an imperative command on one view; returns an optional JSON result.
@_cdecl("pn_bridge_command")
public func pn_bridge_command(
    _ tag: Int64, _ name: UnsafePointer<CChar>?, _ argsJSON: UnsafePointer<CChar>?
) -> UnsafeMutablePointer<CChar>? {
    return PNBridge.onUI {

    PNBridge.ensureMainThread("pn_bridge_command")
    guard let name = name, let record = PNViewRegistry.shared.resolve(tag) else { return nil }
    let args = PNJSON.decodeObject(argsJSON.map { String(cString: $0) })
    let result = record.manager.command(view: record.view, name: String(cString: name), args: args)
    guard let result = result else { return nil }
    return PNBridge.duplicate(PNJSON.encode(result))
    }
}

/// Drive a view's animatable properties (`set` / `start` / `cancel`).
@_cdecl("pn_bridge_animate")
public func pn_bridge_animate(
    _ tag: Int64, _ requestJSON: UnsafePointer<CChar>?
) -> UnsafeMutablePointer<CChar>? {
    return PNBridge.onUI {

    PNBridge.ensureMainThread("pn_bridge_animate")
    let request = PNJSON.decodeObject(requestJSON.map { String(cString: $0) })
    let result = PNAnimator.shared.handle(tag: tag, request: request)
    guard let result = result else { return nil }
    return PNBridge.duplicate(PNJSON.encode(result))
    }
}

/// Call a native module method. The result is `{"ok":..}`, `{"pending":true}`, or an error envelope.
@_cdecl("pn_bridge_call")
public func pn_bridge_call(
    _ module: UnsafePointer<CChar>?, _ method: UnsafePointer<CChar>?, _ argsJSON: UnsafePointer<CChar>?
) -> UnsafeMutablePointer<CChar>? {
    return PNBridge.onUI {

    PNBridge.ensureMainThread("pn_bridge_call")
    guard let module = module, let method = method else {
        return PNBridge.duplicate(PNJSON.encode(["ok": false, "error": "missing module or method"]))
    }
    let envelope = PNJSON.decodeObject(argsJSON.map { String(cString: $0) })
    if String(cString: module) == "Runtime" {
        return PNBridge.duplicate(PNJSON.encode(["ok": true, "value": [
            "protocol": 2, "yoga": "3.2.1", "schema": PNContracts.fingerprint,
            "animation_graph": true, "logical_lists": true, "native_layout": true]]))
    }
    if String(cString: module) == "Layout" {
        return PNBridge.duplicate(PNJSON.encode(["ok": true, "value": PNLayout.compute(envelope["args"] as? [String: Any] ?? [:])]))
    }
    let result = PNModuleDispatcher.shared.call(
        module: String(cString: module), method: String(cString: method), envelope: envelope
    )
    return PNBridge.duplicate(PNJSON.encode(result))
    }
}

/// Free a string previously returned by a `pn_bridge_*` function.
@_cdecl("pn_bridge_free")
public func pn_bridge_free(_ ptr: UnsafeMutablePointer<CChar>?) {
    free(ptr)
}

/// Register the Python callback used for every native -> Python message.
@_cdecl("pn_bridge_set_callback")
public func pn_bridge_set_callback(_ fn: PNCallbackFn?) {
    PNBridge.onUI {

    PNBridge.ensureMainThread("pn_bridge_set_callback")
    PNRegistry.shared.ensureBuiltins()
    PNBridge.shared.setCallback(fn)
    }
}

/// The protocol version this library speaks.
@_cdecl("pn_bridge_protocol_version")
public func pn_bridge_protocol_version() -> Int32 {
    PNProtocolVersion
}
